src/main/java/com/endava/internship/service/UserServiceImpl.java

```diff
@@ -2,68 +2,114 @@
     }
 
     @Override
     public List<User> sortByAgeDescAndNameAsc(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .sorted((a, b) -> {
+                    if (b.getAge().compareTo(a.getAge()) != 0) {
+                        return b.getAge().compareTo(a.getAge());
+                    } else {
+                        return a.getFirstName().compareTo(b.getFirstName());
+                    }
+                })
+                .collect(Collectors.toList());
     }
 
     @Override
     public List<Privilege> getAllDistinctPrivileges(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .map(User::getPrivileges)
+                .flatMap(List::stream)
+                .distinct()
+                .collect(Collectors.toList());
     }
 
     @Override
     public Optional<User> getUpdateUserWithAgeHigherThan(final List<User> users, final int age) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .filter(user -> user.getAge() > age && user.getPrivileges().contains(Privilege.UPDATE))
+                .findFirst();
     }
 
     @Override
     public Map<Integer, List<User>> groupByCountOfPrivileges(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .collect(Collectors.groupingBy(u -> u.getPrivileges().size()));
     }
 
     @Override
     public double getAverageAgeForUsers(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .mapToInt(User::getAge)
+                .average().orElse(-1);
     }
 
     @Override
     public Optional<String> getMostFrequentLastName(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .map(User::getLastName)
+                .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()))
+                .entrySet().stream()
+                .filter(entry -> entry.getValue() >= 2)
+                .reduce((e1, e2) -> {
+                    if (e1.getValue() > e2.getValue()) {
+                        return e1;
+                    } else if (e2.getValue() > e1.getValue()) {
+                        return e2;
+                    } else {
+                        return new SimpleEntry<>(null, e1.getValue());
+                    }
+                }).map(Map.Entry::getKey);
     }
 
     @Override
     public List<User> filterBy(final List<User> users, final Predicate<User>... predicates) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .filter(v -> Arrays.stream(predicates).allMatch(f -> f.test(v)))
```

Comment:
Nice approach, nothing needed to be changed, just for example another approach that can be used:
`.filter(Stream.of(predicates).reduce(x -> true, Predicate::and))`
The idea is to combine all the preficates, but your option also works.